### PR TITLE
Animated swipe back card [with "deck effect"]

### PR DESCRIPTION
**Pull request by webraptor**
**NEW FEATURE: Animated swipe back card [with "deck effect"]**
- Swipe back card is now animated (deck effect)
- Using infinite prop will now result in a never ending deck that cycles automatically [carousel with animation effect] as opposed to previous version when it would go to end and reset to start. 
- Using swipeBackCard with infinite prop means that deck can be swiped bottom to top, not top to bottom
- Ensures only one swipe back animation runs at a time, so no need to check swiper state and block triggers in the parent component 

[excerpt from the update readme file]
## Method

| Name                | type   | description                                        |
| :------------------- | :----- | :------------------------------------------------- |
| swipeBack       | callback                       | swipe back into deck last swiped card |

## Props

| Props                | type   | description                                        | default   |
| :------------------- | :----- | :------------------------------------------------- | :-------- |
| previousCardDefaultPositionX | number | Animation start position oX when card swipes back into deck | -width |
| previousCardDefaultPositionY | number | Animation start position oY when card swipes back into deck | -height |
| stackAnimationFriction | number | spring animation friction (bounciness)                 | 7       |
| stackAnimationTension  | number | spring animation tension (speed)                       | 40      |
| swipeBackCard  | bool | renders swipe back card, in order to animate it                       | false      |

**ALSO FIXES**
https://github.com/alexbrillant/react-native-deck-swiper/issues/171
https://github.com/alexbrillant/react-native-deck-swiper/issues/173

### CODE REVIEW

File: Example/App.js [hunk 1]
```diff
@@ -32,30 +31,11 @@ export default class Exemple extends Component {
     })
   };
 
-  swipeBack = () => {
```
Comment (webraptor) [inline Example/App.js]: This logic was moved to the swiper itself so no one has to repeat this process in any project.

File: Example/package.json [hunk 2]
```diff
@@ -11,7 +11,7 @@
         "prop-types": "^15.5.10",
         "react": "^16.2.0",
         "react-native": "^0.49.5",
-        "react-native-deck-swiper": "^1.5.25"
+        "react-native-deck-swiper": "^1.6.0"
```
Comment (webraptor) [inline Example/package.json]: Added a new feature (animated swipe back, so bumped minor version)

File: Swiper.js [hunk 3]
```diff
@@ -13,38 +13,59 @@ const LABEL_TYPES = {
   TOP: 'top',
   BOTTOM: 'bottom'
 }
+const SWIPE_MULTIPLY_FACTOR = 4.5
+
+const calculateCardIndexes = (firstCardIndex, cards) => {
```
Comment (webraptor) [inline Swiper.js]: Moved _calculateCardIndexes_ and _rebuildStackAnimatedValues_ out of scope so that _getDerivedStateFromProps_ can be used.

**Comment by webraptor** [comment]
@alexbrillant please review it when you have the time. Cheers

File: Swiper.js [hunk 4]
```diff
@@ -364,13 +363,32 @@ class Swiper extends Component {
...
-    Animated.spring(this.state.previousCardY, {
-      toValue: 0,
-      friction: this.props.swipeBackFriction,
-      duration: this.props.swipeBackAnimationDuration
-    }).start(() => {
-      this.decrementCardIndex(cb)
-    })
+    const { swipeBackXYPositions, previousCardX, previousCardY, isSwipingBack } = this.state
+    const { infinite, previousCardDefaultPositionX, previousCardDefaultPositionY } = this.props
+    if (!isSwipingBack && (swipeBackXYPositions.length > 0 || infinite)) {
```
Comment (alexbrillant) [inline Swiper.js]: invert condition, and extract variable :   
```
const canSwipeBack = isSwipingBack || (swipeBackXYPositions.length < 0 && !infinite;
if (!canSwipeBack) {
    return;
}
```

File: Swiper.js [hunk 5]
```diff
@@ -364,13 +363,32 @@ class Swiper extends Component {
...
+      const {x, y} = swipeBackXYPositions.splice(-1, 1)[0] || {x: previousCardDefaultPositionX, y: previousCardDefaultPositionY}
```
Comment (alexbrillant) [inline Swiper.js]: clean code, extract function : 
```
calculateNextPreviousCardPosition = () => {
    const swipeBackPosition = swipeBackXYPositions.splice(-1, 1);
    let x, y;
    if (swipeBackPosition[0]) {
         x = swipeBackPosition.x;
         y = swipeBackPosition.y;
    } else {
         x = previousCardDefaultPositionX;
         y = previousCardDefaultPositionY;
    }
    return { x, y };
}
```

File: Swiper.js [hunk 6]
```diff
@@ -364,13 +363,32 @@ class Swiper extends Component {
...
+      this.setState({isSwipingBack: !isSwipingBack, swipeBackXYPositions}, () => {
+        previousCardX.setValue(x * SWIPE_MULTIPLY_FACTOR)
```
Comment (alexbrillant) [inline Swiper.js]: extract setState callback into another function for the animation : 

```
animatePreviousCard = ({ x, y }) => {
        previousCardX.setValue(x * SWIPE_MULTIPLY_FACTOR);
        previousCardY.setValue(y * SWIPE_MULTIPLY_FACTOR);

        const xAnimation = Animated.spring(this.state.previousCardX, {
            toValue: 0,
            friction: this.props.stackAnimationFriction,
            tension: this.props.stackAnimationTension,
            useNativeDriver: true
          });

        const yAnimation = Animated.spring(this.state.previousCardY, {
            toValue: 0,
            friction: this.props.stackAnimationFriction,
            tension: this.props.stackAnimationTension,
            useNativeDriver: true
          });

        Animated.parallel([xAnimation, yAnimation]).start(() => {
          this.setState({ isSwipingBack: false })
          this.decrementCardIndex(cb)
        })
}
```

**Comment by webraptor** [comment]
@alexbrillant refactored code. Also fixed typo mentioned [here](https://github.com/alexbrillant/react-native-deck-swiper/pull/197)

**Pull request merged**